### DISCUSSION TITLE
How to deal with on/off listeners on service with TypeScript

### DISCUSSION THREAD

**Discussion by danielkcz** [statelyai/xstate#1531] [Q&A] [answered]
I hope this is something that's planned to change in V5. The `service.onTransition` and friends returning service instance is not that convenient as it seems. It should be returning unsubscribe function. Let me show you a little example.

```ts
React.useEffect(() => {
  // extracted the type, but it's horrible to deal with that
  const onTransition: Parameters<typeof service['onTransition']>[0] = (state) => {

  }
  service.onTransition(onTransition)
  return () => {
    service.off(onTransition)
  }
})
```

So ideally it should look like this. Small change, but a huge difference in user code.

```ts
React.useEffect(() => {
  return service.onTransition(state => {
  })
})
```

**Comment by Andarist** [comment]
Keep in mind that React is only an integration point with XState and its APIs should not influence the API decisions for XState itself.

That being said I probably would prefer the proposed change as well. However, this makes patterns like this more awkward to write:
```js
interpret(machine)
  .onTransition(() => {})
  .start()
```

OTOH I feel like `onTransition` should be used really sparingly, so ultimately maybe it doesn't matter much which style is preferred by the core?

**Reply by danielkcz** [reply]
I know that React is not the only concern but honestly even in vanilla if you need to prevent memory leaks it means you have to do something similar I did above (only without useEffect). I think that should be a more important concern that a few extra words in the code.

```js
const service = interpret(machine)
service.onTransition(() => {})
service.start()
```

I am using `onTransition` purely to listen for state change for persistence purposes. I am not aware there would be a better way. Not even considering running action on every event :)

**Reply by Andarist** [reply]
> I know that React is not the only concern but honestly even in vanilla if you need to prevent memory leaks it means you have to do something similar I did above (only without useEffect). 

That's not entirely true - it really depends on how you are using things. For example, this would not require you to even think about that listener and you wouldn't have memory leaks concerns:
```js
const service = interpret(machine)
  .onTransition(() => {})
  .start()

// ...

service.stop()
```

What I'm trying to say is that your argument really only revolves around how one can use it with a particular API (React's API) and it's not a general usability concern. 

> I am using onTransition purely to listen for state change for persistence purposes. I am not aware there would be a better way.

That is one of the good use cases 👍

**Reply by danielkcz** [reply]
Right, did not know that `service.stop()` is that powerful 😅 And I see it's also used in `useMachine` so technically I don't need to care about cleanup then? That's definitely great news. Might be nice to have some summary in docs what `service.stop()` does.

**Reply by davidkpiano** [reply]
> The `service.onTransition` and friends returning service instance is not that convenient as it seems. It should be returning unsubscribe function.

That is what `service.subscribe(...)` is for. It returns an object with the unsubscribe function: `{ unsubscribe: ... }`. The reason it returns a subscription object is because that's part of the [proposed TC39 Observable spec](https://github.com/tc39/proposal-observable).

**Reply by danielkcz** [reply]
@davidkpiano I see, I have no problem with object returned, the annoyance is generally in the need to keep identify of callback around.

So you would generally recommend `service.subscribe` for persistence purposes? It's not clear if `onTransition` does something more or less.

And `service.stop()` won't remove these subscriptions?

**Reply by davidkpiano** [reply]
> the annoyance is generally in the need to keep identify of callback around.

This "annoyance" is not unique to XState (e.g., `window.addEventListener(...)`) 😉 

> So you would generally recommend service.subscribe for persistence purposes? It's not clear if onTransition does something more or less.

They essentially do the same thing.

> And `service.stop()` won't remove these subscriptions?

It will.

**Reply by Andarist** [reply]
> Right, did not know that service.stop() is that powerful 😅 And I see it's also used in useMachine so technically I don't need to care about cleanup then?

Yes and no. Yes - your listeners will be cleaned up, but no - because React might decide to setup the "same" listener multiple times, so if you do not implement cleanup per hook basis then you might end up with multiple same-ish listeners being fired.

The latter problem is not quite solved by `@xstate/react` itself right now so I wouldn't worry about it yet. And, right now, to the best of my knowledge, this can only happen when you are using Fast Refresh.

**Reply by danielkcz** [reply]
@Andarist Thanks for pointing that out, makes total sense. At least there is the `service.subscribe` which makes it much easier.

**Comment by davidkpiano** [comment, marked as answer]
Answer: https://github.com/davidkpiano/xstate/discussions/1531#discussioncomment-91251
